### DISCUSSION TITLE
DDEV simple addon-service routing

### DISCUSSION THREAD

**Discussion by adrianrudnik** [ddev/ddev#5750] [Ideas]
I have a situation where I want to integrate an OIDC-service (imagine Keycloak) that comes with a `.well-known` discovery configuration (think absolute URLs used for configuration, retrieved from the OIDC).

After skimming the documentation there seems to be no simple way to implement a simple traefik to ddev addon-service routing, yet all parameters seem there to support that.

Scenario is:

- I have  keycloak instance., configure `additional_hostnames: ["keycloak.mysite"]`.
- I want to make the keycloak endpoints available through the normal ddev router, port `80` or `443`, have the nice SSL stuff set-up, but want that domain going to keycloak directly.
- As the service requires access by PHP (PHP to OIDC-server for authentication link generation, within LAN) and the browser client (completing the authentication request by visiting that link) there is a disconnect that can only be solved by always going through an external and same-hostnamy configuration, aka https://keycloak.mysite.ddev.site.

What I tried:

- Fiddeling with a custom `docker-compose.keycloak.yaml`, but that only allows hard and conflicting port mappings directly through docker.
- Fiddeling with a custom traefik `mysite.keycloak.yaml`, but that does not prevent the default routes to be registered within the `HostRegexp`, so I need to fight with priority and other stuf, leading to a situation where avoiding DDEV is easier to set-up.
- Fiddeling with NGINX config, but that is also not "nice", because it is the last one in the chain and it could also be any other web server implementation like Apache, and also just reverse-proxying another service, what Traefik could've done in the first place.

What I would like to see:

Optional routing as a one-liner. Here is an example:

```yaml
additional_hostnames:
  - keycloak.mysite:keycloak
 ```

where the `:keycloak` part matches the last part from `ddev-${DDEV_SITENAME}-keycloak` of the ddev-addon service. As the `HTTP_EXPOSE` and `HTTPS_EXPOSE` are also given (or present) the wiring should be pretty clear and uncomplicated. Having `:keycloak` as string also simplifies the struct juggeling within Go. It also helps to identify *disconnected* hostnames in a docker compose-like syntax, so everything that contains a colon must not be routed to the main PHP instance.
```

**Comment by rfay** [comment]
* An add-on can add files to the global or project .ddev directories.
* Traefik configuration can be either global or project, and so an add-on can add to either, https://ddev.readthedocs.io/en/latest/users/extend/traefik-router/#traefik-configuration

I hope that helps and you'll explore this topic! Only a few people explore the amazing things that can be added using nginx. 

The place to start is exploring the traefik configuration to see if you can make it do what you want. Then after that. turn it into an add-on. The add-on is definitely the easy part.
